ui/app/src/state/reducers/preview.ts

```diff
@@ -537,7 +538,22 @@ const reducer = createReducer<GlobalState['preview']>(
     [setHighlightMode.type]: (state, { payload }) => ({
       ...state,
       highlightMode: payload.highlightMode
-    })
+    }),
+    [fetchGlobalPreferencesComplete.type]: (state, { payload }) => {
+      return {
+        ...state,
+        ...(nnou(payload.editMode) && { editMode: payload.editMode === 'true' }),
+        ...(nnou(payload.highlightMode) && { highlightMode: payload.highlightMode })
+      };
+    },
+    [fetchSitePreferencesComplete.type]: (state, { payload }) => {
+      return {
+        ...state,
+        ...(nnou(payload.toolsPanelPage) && {
+          toolsPanelPageStack: [...state.toolsPanelPageStack, JSON.parse(payload.toolsPanelPage)]
```

Comment:
you should check for existence so we don't end up with an null/undefined in the array